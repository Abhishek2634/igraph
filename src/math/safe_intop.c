/*
   IGraph library.
   Copyright (C) 2022 The igraph development team

   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
   02110-1301 USA
*/

#include "math/safe_intop.h"

igraph_error_t igraph_i_safe_add(igraph_integer_t a, igraph_integer_t b, igraph_integer_t *res) {
    IGRAPH_SAFE_ADD(a, b, res);
    return IGRAPH_SUCCESS;
}

igraph_error_t igraph_i_safe_mult(igraph_integer_t a, igraph_integer_t b, igraph_integer_t *res) {
    IGRAPH_SAFE_MULT(a, b, res);
    return IGRAPH_SUCCESS;
}
