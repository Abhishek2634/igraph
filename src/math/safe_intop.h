
#ifndef IGRAPH_MATH_SAFE_INTOP_H
#define IGRAPH_MATH_SAFE_INTOP_H

#include "igraph_types.h"

/* These macros raise an error if the operation would result in an overflow.
 * They must only be used in functions that return an igraph_error_t.
 *
 * This code is based on the recommendation of
 * https://wiki.sei.cmu.edu/confluence/display/c/SEI+CERT+C+Coding+Standard
 */

#define IGRAPH_SAFE_ADD(ia, ib, res) \
    do { \
        igraph_integer_t a = (ia), b = (ib); \
        igraph_integer_t sum; \
        if (((b > 0) && (a > (IGRAPH_INTEGER_MAX - b))) || \
            ((b < 0) && (a < (IGRAPH_INTEGER_MIN - b)))) { \
            IGRAPH_ERRORF("Overflow when adding %"IGRAPH_PRId" and %"IGRAPH_PRId".", IGRAPH_EOVERFLOW, a, b); \
        } \
        sum = a+b; \
        *(res) = sum; \
    } while (0)

#define IGRAPH_SAFE_MULT(ia, ib, res) \
    do { \
        igraph_integer_t a = (ia), b = (ib); \
        igraph_integer_t prod; \
        int err=0; \
        if (a > 0) {  /* a is positive */ \
            if (b > 0) {  /* a and b are positive */ \
                if (a > (IGRAPH_INTEGER_MAX / b)) { \
                    err=1; \
                } \
            } else { /* a positive, b nonpositive */ \
                if (b < (IGRAPH_INTEGER_MIN / a)) { \
                    err=1; \
                } \
            } /* a positive, b nonpositive */ \
        } else { /* a is nonpositive */ \
            if (b > 0) { /* a is nonpositive, b is positive */ \
                if (a < (IGRAPH_INTEGER_MIN / b)) { \
                    err=1; \
                } \
            } else { /* a and b are nonpositive */ \
                if ( (a != 0) && (b < (IGRAPH_INTEGER_MAX / a))) { \
                    err=1; \
                } \
            } /* End if a and b are nonpositive */ \
        } /* End if a is nonpositive */ \
        if (err) { \
            IGRAPH_ERRORF("Overflow when multiplying %"IGRAPH_PRId" and %"IGRAPH_PRId".", IGRAPH_EOVERFLOW, a, b); \
        } \
        prod = a*b; \
        *(res) = prod; \
    } while (0)

#endif
