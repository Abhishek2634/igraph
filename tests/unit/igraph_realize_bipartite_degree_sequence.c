//
// Created by Lara Holm on 1.11.2023.
//
#include <igraph.h>
#include <stdio.h>

#include "test_utilities.h"
#include "math/safe_intop.h"

int main(void) {
    igraph_vector_int_t ds1, ds2;
    igraph_t g;
    igraph_bool_t is_connected;
    igraph_bool_t is_bipartite;
    igraph_bool_t is_simple;
    igraph_vector_int_t degrees;
    igraph_vector_int_t edges;

    // Simple undirected bipartite graph.
    printf("===Simple, connected graph===\n");
    igraph_integer_t deg1[] = {2, 3, 2, 1};
    igraph_integer_t deg2[] = {3, 1, 2, 1, 1};

    igraph_vector_int_init_array(&ds1, deg1, 4);
    igraph_vector_int_init_array(&ds2, deg2, 5);

    igraph_realize_bipartite_degree_sequence(&g, &ds1, &ds2, false);

    igraph_is_simple(&g, &is_simple);
    igraph_is_connected(&g, &is_connected, IGRAPH_STRONG);
    igraph_is_bipartite(&g, &is_bipartite, NULL);

    printf("Edges: ");
    igraph_vector_int_init(&edges, 0);
    igraph_get_edgelist(&g, &edges, false);
    igraph_vector_int_print(&edges);

    printf("Degrees: ");
    igraph_vector_int_init(&degrees, 0);
    igraph_degree(&g, &degrees, igraph_vss_all(), IGRAPH_ALL, false);
    igraph_vector_int_print(&degrees);

    printf("vcount: %d\n", (int)igraph_vcount(&g));
    printf("Simple: %d\n", is_simple);
    printf("Connected: %d\n", is_connected);
    printf("Bipartite: %d\n", is_bipartite);

    igraph_vector_int_destroy(&degrees);
    igraph_vector_int_destroy(&edges);
    igraph_vector_int_destroy(&ds1);
    igraph_vector_int_destroy(&ds2);
    igraph_destroy(&g);

    // undirected bipartite multigraph.
    igraph_bool_t has_multi;
    printf("===Connected multigraph===\n");

    igraph_integer_t deg1m[] = {2, 3, 1};
    igraph_integer_t deg2m[]= {4, 2};

    igraph_vector_int_init_array(&ds1, deg1m, 3);
    igraph_vector_int_init_array(&ds2, deg2m, 2);

    igraph_realize_bipartite_degree_sequence(&g, &ds1, &ds2, true);

    printf("Edges: ");
    igraph_vector_int_init(&edges, 0);
    igraph_get_edgelist(&g, &edges, false);
    igraph_vector_int_print(&edges);

    printf("Degrees: ");
    igraph_vector_int_init(&degrees, 0);
    igraph_degree(&g, &degrees, igraph_vss_all(), IGRAPH_ALL, false);
    igraph_vector_int_print(&degrees);

    igraph_has_multiple(&g, &has_multi);
    igraph_is_connected(&g, &is_connected, IGRAPH_STRONG);
    igraph_is_bipartite(&g, &is_bipartite, NULL);

    printf("vcount: %d\n", (int)igraph_vcount(&g));
    printf("Has multiedges: %d\n", has_multi);
    printf("Connected: %d\n", is_connected);
    printf("Bipartite: %d\n", is_bipartite);

    igraph_vector_int_destroy(&degrees);
    igraph_vector_int_destroy(&edges);
    igraph_vector_int_destroy(&ds1);
    igraph_vector_int_destroy(&ds2);
    igraph_destroy(&g);

    VERIFY_FINALLY_STACK();
    return 0;
}
