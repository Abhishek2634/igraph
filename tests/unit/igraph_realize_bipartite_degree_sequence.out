===Simple, connected graph===
Edges: 3 4 1 8 1 7 1 4 0 5 0 6 2 4 2 6
Degrees: 2 3 2 1 3 1 2 1 1
vcount: 9
Simple: 1
Connected: 1
Bipartite: 1
===Connected multigraph===
Edges: 2 3 0 3 0 3 1 3 1 4 1 4
Degrees: 2 3 1 4 2
vcount: 5
Has multiedges: 1
Connected: 1
Bipartite: 1
